---
layout: post
title:  "Consensus and Profile"
date:   2025-07-26
---

[This](https://rosalind.info/problems/cons/) problem asks:

> **Given**: A collection of at most 10 DNA strings of equal length (at most 1 kbp) in FASTA format.

> **Return**: A consensus string and profile matrix for the collection. 

<!--Break-->

# Restate the problem
Reading required for me to restate the problem:
1. [DNA sequence alignment, specifically motif finding](https://en.wikipedia.org/wiki/Sequence_alignment#Motif_finding)
2. [DNA consensus sequences](https://en.wikipedia.org/wiki/Consensus_sequence)
3. [Most likely common ancestor](https://en.wikipedia.org/wiki/Most_recent_common_ancestor)

I'm going to get a set of DNA string **of equal length**. I need to return two items:
1. a string with the most common symbol shown at each position
2. a matrix with four rows (A, C, G, and T) and as many columns as the DNA strings are long. In each position in the matrix, include the count of the specific symbol at that location.

# Solution steps
Biopython has a module called "[motifs](https://biopython.org/docs/latest/Tutorial/chapter_motifs.html)" that includes both a consensus function and a count matrix function, plus it comes with a tutorial!

This solution requires a manual edit of the solution file to remove the column headers from the matrix section.

# Python concepts
string replace()
list methods
[method chains](https://www.geeksforgeeks.org/python/method-chaining-in-python/)

# Bioinformatics concepts
[Most likely common ancestor](https://en.wikipedia.org/wiki/Most_recent_common_ancestor)


# Problem solving concepts
When it's good enough.


