---
layout: post
title:  "Open Reading Frames"
date:   2025-07-27
---

[This](https://rosalind.info/problems/orf/) problem asks:

> **Given**: A DNA string s of length at most 1 kbp in FASTA format.

> **Return**: Every distinct candidate protein string that can be translated from ORFs of s.

<!--Break-->

# Required reading:
1. [Open reading frames](https://en.wikipedia.org/wiki/Open_reading_frame)
2. [Clarifying definition of ORFs](https://www.cell.com/trends/genetics/fulltext/S0168-9525(17)30229-9?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0168952517302299%3Fshowall%3Dtrue)
3. [Exon and coding region](https://en.wikipedia.org/wiki/Coding_region)
4. [DNA codon table _reread_](https://en.wikipedia.org/wiki/Genetic_code#DNA_codon_table)
5. 

# Restate the problem
I'm going to get a single DNA string. I need to identify the regions between [start codons](https://en.wikipedia.org/wiki/Start_codon) and [stop codons](https://en.wikipedia.org/wiki/Stop_codon). Then, I need to decode those regions using the [DNA codon table](https://rosalind.info/glossary/dna-codon-table/).

# Solution steps
I found [this section of the Biopython cookbook](https://biopython.org/docs/latest/Tutorial/chapter_cookbook.html#identifying-open-reading-frames) that details how to build an ORF finder with Python.
I found the [PyPi project ORFFinder](https://pypi.org/project/orffinder/) that includes methods for identifying ORFs in DNA strings.
The orffinder method get ORFProteins() returns a partial solution, but does not include all the possible proteins that can be decoded from the DNA sequences in the ORF. I read the [source code for orffinder](https://github.com/Chokyotager/ORFFinder/blob/main/src/orffinder/orffinder.py), but didn't see a way to customize it to include all possible proteins.



# Python concepts
pip, installing new libraries

# Bioinformatics concepts


