---
layout: post
title:  "Mendel's First Law"
date:   2025-07-18
---

[This](https://rosalind.info/problems/iprb/) problem asks:

> **Given**: Three positive integers _k_, _m_, and _n_, representing a population containing k+m+n organisms: k individuals are homozygous dominant for a factor, m are heterozygous, and n are homozygous recessive.

> **Return**: The probability that two randomly selected mating organisms will produce an individual possessing a dominant allele (and thus displaying the dominant phenotype). Assume that any two organisms can mate.

# First impression
The system gets two [alleles](https://rosalind.info/glossary/allele/) from each member in _k_, one from each member in _m_, and none from each member in _n_.

They want to know the likelihood that a random child from this population will have at least one allele.

There is a hint, which I'll keep in mind:
> Consider simulating inheritance on a number of small test cases in order to check your solution.

# Solution steps
First, I read [this LibreText Biology textbook section](https://bio.libretexts.org/Bookshelves/Genetics/Online_Open_Genetics_(Nickle_and_Barrette-Ng)/03%3A_Genetic_Analysis_of_Single_Genes/3.01%3A__Mendels_First_Law) about Mendel's First Law, which was helpful, but didn't get into calculating probabilities.

[This article](https://mathsathome.com/probability-tree-diagrams/) on probability trees was nice, but didn't move my solution forward.

I had the idea to run a million random trials to estimate the probability of the offspring of two random parent having a dominant allele,
but I didn't try to code that because it didn't seem like a "real" solution, and I was afraid that a million trails wouldn't be
enough to get the right answer.

Instead, I wanted to count each one of the possible combinations of parents one time. One helpful search term was "[Combinatorics](https://en.wikipedia.org/wiki/Combinatorics)".
That lead me to the [itertools library](https://docs.python.org/3/library/itertools.html) in Python, and specifically [combinations](https://docs.python.org/3/library/itertools.html#itertools.combinations).
However, I wasn't able to make any progress on a solution, even though this iterator looks perfect for the job.

Then, I read [this](https://susannahgo.wordpress.com/wp-content/uploads/2015/11/rosalind-iprb.pdf) "solution", which seems to prove that there's a direct mathematical
way to calculate this probability, but I don't understand that math, so it was no help to me. [This person](https://saradoesbioinformatics.blogspot.com/2016/06/mendels-first-law.html) comes to the same conclusion, but again, I don't understand that math.

After some struggling, I had this idea:
>   population = k * ["AA"] + m * ["Aa"] + n * ["aa"]

This generates a set where each item in the set being the genome of an individual in the population.

Then I could use [itertools.combinations()](https://docs.python.org/3/library/itertools.html#itertools.combinations)
to make a set of all the possible pairs of parents in the population.

> possible_pairs = sorted(itertools.combinations(population, 2))

Then the only math I needed was:

> If your parents are ('AA', 'AA'), ('AA', 'Aa'), or ('AA, 'aa'), you have a 100% chance of getting an allele.
> If your parents are ('Aa', 'Aa'), you have a 75% chance of getting an allele.
> If your parents are ('Aa', 'aa'), you have a 50% chance of getting an allele.

The core of my [code](https://github.com/rmbryan71/rosalind/blob/main/solution-code/iprb.py) looks like this:
```aiignore
        for x in range(trial_size):
            parents = random.choice(possible_pairs)
            if parents in (('AA', 'AA'), ('AA', 'Aa'), ('AA', 'aa')):
                count += 1
            if parents == ('Aa', 'Aa'):
                count += .75
            if parents == ('Aa', 'aa'):
                count += .50
        return str(round((count/trial_size), 3))
```

# Conclusion
To be written